staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/stdlib.go

```diff
@@ -0,0 +1,616 @@
+		next := it.Next()
+		nextCmp, ok := next.(traits.Comparer)
+		if !ok {
+			return types.NoSuchOverloadErr()
+		}
+		if prev != nil {
+			cmp := prev.Compare(next)
+			if cmp == types.IntOne {
+				return types.False
+			}
+		}
+		prev = nextCmp
+	}
+	return types.True
+}
+
+func sum(init func() ref.Val) functions.UnaryOp {
+	return func(val ref.Val) ref.Val {
+		acc, ok := init().(traits.Adder)
+		if !ok {
+			// Should never happen since all passed in init values are valid
+			return types.NoSuchOverloadErr()
+		}
+		iterable, ok := val.(traits.Iterable)
+		if !ok {
+			return types.NoSuchOverloadErr()
+		}
+		for it := iterable.Iterator(); it.HasNext() == types.True; {
+			next := it.Next()
+			nextAdder, ok := next.(traits.Adder)
+			if !ok {
+				// Should never happen for type checked CEL programs
+				return types.NoSuchOverloadErr()
+			}
+			if acc != nil {
+				sum, ok := acc.Add(next).(traits.Adder)
+				if !ok {
+					// Should never happen for type checked CEL programs
+					return types.NoSuchOverloadErr()
+				}
+				acc = sum
+			} else {
+				acc = nextAdder
+			}
+		}
+		return acc.(ref.Val)
+	}
+}
+
+func min() functions.UnaryOp {
+	return cmp("min", types.IntOne)
+}
+
+func max() functions.UnaryOp {
+	return cmp("max", types.IntNegOne)
+}
+
+func cmp(opName string, opPreferCmpResult ref.Val) functions.UnaryOp {
+	return func(val ref.Val) ref.Val {
+		var result traits.Comparer
+		iterable, ok := val.(traits.Iterable)
+		if !ok {
+			return types.NoSuchOverloadErr()
+		}
+		for it := iterable.Iterator(); it.HasNext() == types.True; {
+			next := it.Next()
+			nextCmp, ok := next.(traits.Comparer)
+			if !ok {
+				// Should never happen for type checked CEL programs
+				return types.NoSuchOverloadErr()
+			}
+			if result == nil {
+				result = nextCmp
+			} else {
+				cmp := result.Compare(next)
+				if cmp == opPreferCmpResult {
+					result = nextCmp
+				}
+			}
+		}
+		if result == nil {
+			return types.NewErr("%s called on empty list", opName)
+		}
+		return result.(ref.Val)
+	}
+}
+
+func indexOf(list ref.Val, item ref.Val) ref.Val {
+	lister, ok := list.(traits.Lister)
+	if !ok {
+		return types.NoSuchOverloadErr()
+	}
+	sz := lister.Size().Value().(int64)
+	for i := int64(0); i < sz; i++ {
+		if lister.Get(types.Int(i)).Equal(item) == types.True {
+			return types.Int(i)
+		}
+	}
+	return types.Int(-1)
+}
+
+func lastIndexOf(list ref.Val, item ref.Val) ref.Val {
+	lister, ok := list.(traits.Lister)
+	if !ok {
+		return types.NoSuchOverloadErr()
+	}
+	sz := lister.Size().Value().(int64)
+	for i := sz - 1; i >= 0; i-- {
+		if lister.Get(types.Int(i)).Equal(item) == types.True {
+			return types.Int(i)
+		}
+	}
+	return types.Int(-1)
+}
+
+func find(strVal ref.Val, regexVal ref.Val) ref.Val {
+	str, ok := strVal.Value().(string)
+	if !ok {
+		return types.NoSuchOverloadErr()
+	}
+	regex, ok := regexVal.Value().(string)
+	if !ok {
+		return types.NoSuchOverloadErr()
+	}
+	re, err := regexp.Compile(regex)
```

Comment:
is it possible to do the compile of static regexes once at cel compile time rather than on every evaluation?

Reply:
I REALLY want this. I'm going to poke around a bit with CEL and see what the options are.

Reply:
You can do this with a decorator that looks at an `InterpretableCall` and swaps the implementation of the call node with one that memoizes the regex pattern. You'll need a different impl for find and findAll that reads the pattern from a stored variable.

Reply:
I've opened https://github.com/google/cel-go/issues/496 to track.  Compiling the regex at CEL compile time also means we can report any regex compilation errors at compile time, which is a nice type-safety benefit.